### ISSUE TITLE
High memory consumption at F.affine_grid()

### ISSUE THREAD

**Issue by FrostbiteXSW** [bdsaglam/torch-scae#1]
Hi, thanks for you implementation. I've tried this model on my pc with GTX 1050 Ti. Unfortunately, this model consumed huge amount of memory at this block and soon my GPU memory ran out:
```
            td_feature = part_enc_res.feature
            if td_feature is not None:
                td_feature = td_feature.repeat(n_obj_caps, 1, 1)

            td_templates = self.template_generator(feature=td_feature).templates
            td_pose = res.vote.view(-1, *res.vote.shape[2:])
            td_presence = res.vote_presence.view(-1, *res.vote_presence.shape[2:]) \
                          * part_enc_res.presence.repeat(n_obj_caps, 1)
            res.top_down_per_caps_rec = self.part_decoder(
                templates=td_templates,
                pose=td_pose,
                presence=td_presence)
            del td_feature, td_templates, td_pose, td_presence
```
This happened after this line in part_decoder.py:
```
            affine_grids = F.affine_grid(affine_matrices, target_size)
```
It seems like some variables aren't released properly. Even if I set the batch_size value to 4, GPU memory will still run out. Just a matter of time.
However, the offical tensorflow implementation at https://github.com/akosiorek/stacked_capsule_autoencoders and other pytorch implementation at https://github.com/phanideepgampa/stacked-capsule-networks can run well (while that pytorch implementation has some problems). The offcial one also gains faster loss decreasing speed. I wonder if there is any solution to this problem?

**Comment by bdsaglam** [comment]
Hi, thanks for reporting the problem. 
I've noticed memory problem as well and recently added manual garbage collection after each batch (commit 8211d641). Did you run your experiment before or after that commit? 
I have spotted a bug in tiling tensors and an unnecessary tiling of features. I fixed them, hope it solves the problem. 
If not, can you try training the model without that block since it is not included in loss calculation? 
I have run the model on Colab for only 10 epochs with batch size of 64, but did not train it further. 
I have seen the other PyTorch implementation but haven't trained it since it didn't seem good to me. I mostly kept my implementation same with official TensorFlow implementation except Set Transformer which I made it similar to the official implementation of Set Transformer paper. There might be a mistake causing the difference you mentioned. 
Unfortunately, I currently don't have time and resources for investigating these problems in detail immediately, but I will probably be able check them before July.

**Comment by FrostbiteXSW** [comment]
Thanks for reply. I'm using the version with del and gc.collect(). But in my experiment, it has no effect on this problem. Also I've run the model with this block removed for 10 epochs, and it finished successfully. Maybe I will train it for more epochs in the future. If there is any other problem, I will open a new issue to let you know.

**Comment by bdsaglam** [comment]
Hi again, I have made some changes.

- Added a flag argument, ```reconstruct_alternatives```,  to the SCAE class for optional reconstructions. Without computing them, I was able to train for batch size of 512 with 10 GB of GPU memory. 
- Changed the Set Transformer implementation to be same with original implementation.
- Made RMSProp as default optimizer with same parameters. Also, added Adam and RAdam optimizer as options.

Hope these improve the convergence behavior.
